### PR TITLE
Support for up/down arrows as lcd control buttons

### PR DESCRIPTION
**Pull request by GilesBathgate**
### Description

Adds an intuitive keyboard mapping for the up/down cursor keys, to control the lcd

### Behaviour/ Breaking changes

~Only implemented for Main window not the GL viewer.~

### Have you tested the changes?

Yes - No unit tests (yet)

### Other

~Basically a PoC. Let me know how to approach code centralisation for the GL view if the feature is wanted.~

### CODE REVIEW

**Comment by vintagepc** [comment]
Hmm... I need to think about this some more. I think the feature is good, but since the "special" keys are an entirely separate callback channel to handle, but for uniformity I think it might be best to mirror the way we do it for the standard keys where we route the key handling directly in to the KeyController so that any client can "subscribe" to special keys too.  (Bonus - this also makes it testable using the scripting framework)

Not sure what makes more sense currently - we could either modify the current OnKeyPressed() call chain to be 16 bit instead of 8, but that's a lot of churn. 

Perhaps it makes more sense to just use a mask on the `uchar` MSB (since we don't use keys above ASCII 127 anyway)  to signify this is a key from the "special" key set.

File: MK404.cpp [hunk 1]
```diff
@@ -140,6 +140,16 @@ void KeyCB(unsigned char key, int /*x*/, int y)
 	}
 }
 
+void SpecialKeyCB(int key, int x, int y)
+{
+	if (key==0x65) { // Up
```
Comment (vintagepc) [inline MK404.cpp:145]: There are `GLUT_KEY_UP` and `GLUT_KEY_DOWN` #defines that are probably better used here rather than magic numbers

**Comment by GilesBathgate** [comment]
> Perhaps it makes more sense to just use a mask on the `uchar` MSB in OnKeyPressed (since we don't use keys above ASCII 127 anyway) to signify this is a key from the "special" key set.

You could probably route the two static callbacks through to a single `OnKeyPressed(char key, int keycode, int x, int y)`.
It's much like the way in which the current implementation doesn't use x and y in most cases. Although it's not a strictly OOP design, it saves a lot of duplication potentially.

**Comment by vintagepc** [comment]
I think we're saying the same thing in different ways. 

E.g. add the following to KeyController.h:

**Comment by vintagepc** [comment]
I think we're saying the same thing in different ways. 

E.g. add the following to KeyController.h:
`static constexpr uint8_t SPECIAL_KEY_MASK = 0x80`
```
static inline void GLSpecialKeyReceiver(unsigned char key, int /*x*/, int /*y*/) { KeyController::GetController().OnKeyPressed(key |  SPECIAL_KEY_MASK); };
```

and then any existing client can subscribe to special keys this way; for example the encoder just needs to indicate it wishes these and add the mask as an appropriate entry in the OnKeyPressed() switch handler.

**Comment by GilesBathgate** [comment]
> and then any existing client can subscribe to special keys this way; for example the encoder just needs to indicate it wishes these and add the mask as an appropriate entry in the OnKeyPressed() switch handler.

I guess, it just seems a bit of a pain for the client/subscriber to have to mask off the bits for the sake of putting both data in the same byte of the parameter. Does it lend itself to Utf-8 based keycodes if i18n keyboards are used in future? 

It does mean fewer changes though.

**Comment by vintagepc** [comment]
I figure for two keys the mask (while a bit ugly) is faster and less code churn than refactoring all of the function definitions and declarations, (and adding a method for clients to subscribe to special keys) but I'm not opposed to that in general since it is the "better" approach from a legibility perspective. But doing this does also open the door for things like using the F-keys for shortcuts, at which point it does tip the balance in favor of doing it "right".

**Comment by GilesBathgate** [comment]
> static constexpr uint8_t SPECIAL_KEY_MASK = 0x80

Something like the latest commit, but it doesn't work. Spot my mistake ;)

**Comment by GilesBathgate** [comment]
@vintagepc Also what is up with the include sort order, its correct surely?

File: parts/components/RotaryEncoder.cpp [hunk 2]
```diff
@@ -21,11 +21,12 @@
 	along with MK404.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-
 #include "RotaryEncoder.h"
+#include "KeyController.h"
```
Comment (vintagepc) [inline parts/components/RotaryEncoder.cpp:25]: Clang-tidy expects them to be sorted alphabetically, you have `K` after `R`

Comment (GilesBathgate) [reply parts/components/RotaryEncoder.cpp:24]: I had `K` before `R` and it didn't like that either.

**Comment by GilesBathgate** [comment]
@vintagepc Need to do something about the following:

```
$ ./MK404 --keys 
Available Key Controls:
	Enter:	Pushes and releases the encoder button 
	+:	Start VCD trace 
	-:	Stop VCD trace 
	1:	Toggles LCD Colour scheme 
	S:	Take a snapshot of the LCD 
	c:	Toggle SD card mount/unmount 
	f:	Toggles IR sensor Filament presence 
	h:	Pushes and long-holds the encoder button. 
	m:	Mutes buzzer audio tones 
	p:	Triggers Power Panic line 
	q:	Shuts down the board and exits 
	r:	Resets the AVR/board 
	s:	Twists encoder CW 
	t:	Triggers a factory reset (reset + hold encoder 
	w:	Twists encoder CCW 
	y:	Toggles the steel sheet on the heatbed 
	z:	Pauses/resumes AVR execution 
	�:	Twists encoder CCW 
	�:	Twists encoder CW
```
i.e the last two lines.

**Comment by vintagepc** [comment]
clang-tidy expects them to be in alphabetical order, you have K below 

> @vintagepc Need to do something about the following:
> 
> ```
> $ ./MK404 --keys 
> Available Key Controls:
> 	Enter:	Pushes and releases the encoder button 
> 	+:	Start VCD trace 
> 	-:	Stop VCD trace 
> 	1:	Toggles LCD Colour scheme 
> 	S:	Take a snapshot of the LCD 
> 	c:	Toggle SD card mount/unmount 
> 	f:	Toggles IR sensor Filament presence 
> 	h:	Pushes and long-holds the encoder button. 
> 	m:	Mutes buzzer audio tones 
> 	p:	Triggers Power Panic line 
> 	q:	Shuts down the board and exits 
> 	r:	Resets the AVR/board 
> 	s:	Twists encoder CW 
> 	t:	Triggers a factory reset (reset + hold encoder 
> 	w:	Twists encoder CCW 
> 	y:	Toggles the steel sheet on the heatbed 
> 	z:	Pauses/resumes AVR execution 
> 	�:	Twists encoder CCW 
> 	�:	Twists encoder CW
> ```
> 
> i.e the last two lines.

We'll need to add entries for the arrow keys to:

`KeyController::ProcessAction()` (for scripting to be able to do e.g. "KeyPress(up)"
and `KeyController::PutNIceKeyName()` to have the `--help` output print nicely.

**Comment by GilesBathgate** [comment]
@vintagepc Still doesn't like the sort order.

**Comment by vintagepc** [comment]
ugh, github lost part of my comment. It was supposed to say that the GL include needs to be just before `<algorithm>`. 

So the correct order is:
```
#include "KeyController.h"
#include "RotaryEncoder.h"
#include "TelemetryHost.h"
#include "gsl-lite.hpp"
#include <GL/glut.h>
#include <algorithm>         // for copy
#include <iostream>
```

Overall this is starting to look pretty good. If you wouldn't also mind adding some test cases for the arrow keys to scripts/tests/test_encoder,txt and scripts/tests/test_keys_lite_gfx.txt that'd be perfect. But I'm also happy to do that as well since the test framework can be a little peculiar to understand :) That should also get the codecov bot to shut up about the diff not being covered :+1:

**Comment by GilesBathgate** [comment]
> Overall this is starting to look pretty good. If you wouldn't also mind adding some test cases for the arrow keys to scripts/tests/test_encoder,txt and scripts/tests/test_keys_lite_gfx.txt that'd be perfect. But I'm also happy to do that as well since the test framework can be a little peculiar to understand :) That should also get the codecov bot to shut up about the diff not being covered +1
> 
> Don't worry too much about trying to get the snapshots, because we use SWGL rendering on the build runner I basically have to pull the "correct" test images from the build and drop those in since locally rendered images with hardware GL won't pass the equivalence test.

Sounds tricky, I'd rather give it a miss ;)

**Comment by GilesBathgate** [comment]
> We'll need to add entries for the arrow keys to:
> 
> `KeyController::ProcessAction()` (for scripting to be able to do e.g. "KeyPress(up)" and `KeyController::PutNIceKeyName()` to have the `--help` output print nicely.

Not really sure what's needed for that, I've just suppressed them by setting the descriptions blank.

**Comment by vintagepc** [comment]
No worries - thanks for your submission :) I Should have some time later today or tomorrow to add the test coverage and then I'll get it merged.

File: parts/KeyController.cpp [hunk 3]
```diff
@@ -78,6 +87,12 @@ void KeyController::PutNiceKeyName(unsigned char key)
 		case 0xd:
 			std::cout << "Enter";
 			break;
+		case GLUT_KEY_UP | SPECIAL_KEY_MASK:
+			std::cout << "Up Arrow";
+			break;
+		case GLUT_KEY_DOWN | SPECIAL_KEY_MASK:
+			std::cout << "DOwn Arrow";
```
Comment (GilesBathgate) [inline parts/KeyController.cpp:94]: Typo?

Comment (vintagepc) [reply parts/KeyController.cpp:94]: D'oh... one of my shift keys is not well molded and sometimes gets stuck for an extra keystroke. One of these days I'll get bothered enough to fix that XD....

**Pull request merged**